File: take_u_forward/recursion/4_Problems on Functional Recursion.java
//Q1. Reverse an array using recursion

import java.util.Scanner;
class Main {
    
    public static void fun(int i, int arr[],int n){
        if(i>=n/2) return;
        int temp=arr[i];
        arr[i]=arr[n-i-1];
        arr[n-i-1]=temp;
        fun(i+1,arr,n);
    }
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter the size of the array:\t");
        int n=sc.nextInt();
        System.out.print("Enter the element in the array:\t");
        int[] arr=new int[n];
        for(int i=0;i<n;i++){
            arr[i]=sc.nextInt();
        }
        fun(0,arr,n);
        System.out.print("Array element after swap:\t");
        for(int i=0;i<n;i++){
            System.out.print(arr[i]+" ");
        }
    }
}


//Q1. Check if string is palindrome or not using recursion

import java.util.Scanner;
class Main {
    public static boolean fun(int i, String s){
        if(i>=s.length()/2) return true;
        if(s.charAt(i)!=s.charAt(s.length()-i-1)) return false;
        return fun(i+1,s);
    }
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter a string:\t");
        String s=sc.next();
        System.out.println(fun(0,s));
    }
}

File: take_u_forward/recursion/4_Problems on Functional Recursion.cpp
//Q1. Reverse an array using recursion
#include <iostream>
using namespace std;

void fun(int i, int arr[], int n){
    if(i>=n/2) return;
    swap(arr[i],arr[n-i-1]);
    fun(i+1,arr,n);
}

int main()
{
    int n;
    cout<<"Enter the array size:\t";
    cin>>n;
    int arr[n];
    cout<<"Enter the element in the array:\t";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    
    fun(0,arr,n);
    
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }

    return 0;
}

//Q1. Check if string is palindrome or not using recursion

#include <iostream>
using namespace std;
bool fun(int i, string &s){
    if(i>=s.size()/2) return true;
    if(s[i]!=s[s.size()-i-1]) return false;
    return fun(i+1,s);
}

int main() {
    // Write C++ code here
    string s;
    cout << "Enter the string:\t";
    cin >> s;
    cout << fun(0,s);

    return 0;
}